src/lucky/routeable.cr

```diff
@@ -115,4 +145,49 @@ module Lucky::Routeable
       \{% raise "Use `route` instead of `with` if the action doesn't need params" %}
     end
   end
+
+  macro included
+    PARAM_DECLARATIONS = [] of Crystal::Macros::TypeDeclaration
+
+    macro inherited
+      PARAM_DECLARATIONS = [] of Crystal::Macros::TypeDeclaration
+    end
+  end
+
+  macro param(type_declaration)
+    {% PARAM_DECLARATIONS << type_declaration %}
+
+    def {{ type_declaration.var }} : {{ type_declaration.type }}
+      {% if type_declaration.type.class_name != "Union" %}
+        val = params.get(:{{ type_declaration.var.id }})
+        if val.nil?
+          return {{ type_declaration.value || nil }} ||
+            raise Lucky::Exceptions::MissingParam.new("{{ type_declaration.var.id }}")
+        end
+        result = {{ type_declaration.type }}::Lucky.parse(val)
+
+        if result.is_a? {{ type_declaration.type }}::Lucky::SuccessfulCast
+          return result.value
+        end
+
+        raise Lucky::Exceptions::InvalidParam.new(
+          "{{ type_declaration.var.id }}",
+          val.to_s,
+          "{{ type_declaration.type }}"
+        )
+      {% else %}
+        val = params.get(:{{ type_declaration.var.id }})
+        if val.nil?
+          {{ type_declaration.value || nil }}
+        else
+          result = {{ type_declaration.type.types.first }}::Lucky.parse(val)
+          if result.is_a? {{ type_declaration.type.types.first }}::Lucky::SuccessfulCast
+            result.value
+          else
+            nil
+          end
+        end
+      {% end %}
```

Comment:
Instead of having an if/else you could use some macro variables to keep a lot of the code the same:

```crystal
{% optional = type_declaration.type.is_a?(Union) %}
# Uses get! if required. This will raise MissingParam if missing
val = params.get{% unless optional %}!{% end %}(:{{ type_declaration.var.id }})

result = {{ type_declaration.type }}::Lucky.parse(val)

# Use if/else instead of early return to make things clear
if result.is_a? {{ type_declaration.type }}::Lucky::SuccessfulCast
  result.value
else
  # Use named args to make it clear what each arg is
  raise Lucky::Exceptions::InvalidParam.new(
    param_name: "{{ type_declaration.var.id }}",
    param_value: val.to_s,
    param_type: "{{ type_declaration.type }}"
  )
end
```

This also incorporates a few other changes I suggested earlier, like always raising if a cast fails

Reply:
I took a lot of inspiration from your comment, but it's not quite that simple because we have to do a default check for required params as well. So we either do `get!` and catch the exception, or do a nil check. I opted for the second one for now, not that I particularly like it^^